### PR TITLE
Fix floating point issues (Issue #538)

### LINKED ISSUE #538: Floating point issue when choosing holdout set train size
This line of [code](https://github.com/automl/auto-sklearn/blob/master/autosklearn/evaluation/train_evaluator.py#L549) will introduce floating point issue, e.g., when `train_size=0.7` will yield `0.30000000000000004` instead of `0.3`, thus causing the following error message
 
```
[ERROR] [2018-08-27 16:34:59,972:AutoML(1):cb3d3adc8ee73def12eda5903e98a85d] Error creating dummy predictions: {'traceback': 'Traceback (most recent call last):\n  File "/root/anaconda3/lib/python3.6/site-packages/autosklearn/evaluation/__init__.py", line 30, in fit_predict_try_except_decorator\n    return ta(queue=queue, **kwargs)\n  File "/root/anaconda3/lib/python3.6/site-packages/autosklearn/evaluation/train_evaluator.py", line 644, in eval_holdout\n    init_params=init_params\n  File "/root/anaconda3/lib/python3.6/site-packages/autosklearn/evaluation/train_evaluator.py", line 93, in __init__\n    self.cv = self.get_splitter(self.datamanager)\n  File "/root/anaconda3/lib/python3.6/site-packages/autosklearn/evaluation/train_evaluator.py", line 568, in get_splitter\n    raise e\n  File "/root/anaconda3/lib/python3.6/site-packages/autosklearn/evaluation/train_evaluator.py", line 562, in get_splitter\n    next(test_cv.split(y, y))\n  File "/root/anaconda3/lib/python3.6/site-packages/sklearn/model_selection/_split.py", line 1204, in split\n    for train, test in self._iter_indices(X, y, groups):\n  File "/root/anaconda3/lib/python3.6/site-packages/sklearn/model_selection/_split.py", line 1534, in _iter_indices\n    self.train_size)\n  File "/root/anaconda3/lib/python3.6/site-packages/sklearn/model_selection/_split.py", line 1710, in _validate_shuffle_split\n    \'train_size.\' % (n_train + n_test, n_samples))\nValueError: The sum of train_size and test_size = 17221, should be smaller than the number of samples 17220. Reduce test_size and/or train_size.\n', 'error': "ValueError('The sum of train_size and test_size = 17221, should be smaller than the number of samples 17220. Reduce test_size and/or train_size.',)", 'configuration_origin': 'DUMMY'} 
[ERROR] [2018-08-27 16:35:01,245:__main__] Unexpected Exception
Traceback (most recent call last):
  File "/var/www/html/MC_PY/mc_modeling/modeling.py", line 26, in main
    alg.fit()
  File "/var/www/html/MC_PY/mc_modules/mc_automl/base.py", line 456, in fit
    self.train()
  File "/var/www/html/MC_PY/mc_modules/mc_automl/classification.py", line 60, in train
    self.automl.fit(self.X_train.copy(), self.y_train.copy(), metric=self.metric, feat_type=self.feature_type)
  File "/root/anaconda3/lib/python3.6/site-packages/autosklearn/estimators.py", line 466, in fit
    dataset_name=dataset_name,
  File "/root/anaconda3/lib/python3.6/site-packages/autosklearn/estimators.py", line 248, in fit
    self._automl.fit(*args, **kwargs)
  File "/root/anaconda3/lib/python3.6/site-packages/autosklearn/automl.py", line 954, in fit
    only_return_configuration_space=only_return_configuration_space,
  File "/root/anaconda3/lib/python3.6/site-packages/autosklearn/automl.py", line 199, in fit
    only_return_configuration_space,
  File "/root/anaconda3/lib/python3.6/site-packages/autosklearn/automl.py", line 461, in _fit
    _proc_smac.run_smbo()
  File "/root/anaconda3/lib/python3.6/site-packages/autosklearn/smbo.py", line 501, in run_smbo
    smac.optimize()
  File "/root/anaconda3/lib/python3.6/site-packages/smac/facade/smac_facade.py", line 400, in optimize
    incumbent = self.solver.run()
  File "/root/anaconda3/lib/python3.6/site-packages/smac/optimizer/smbo.py", line 180, in run
    challengers = self.choose_next(X, Y)
  File "/root/anaconda3/lib/python3.6/site-packages/smac/optimizer/smbo.py", line 247, in choose_next
    incumbent_value = self.runhistory.get_cost(self.incumbent)
  File "/root/anaconda3/lib/python3.6/site-packages/smac/runhistory/runhistory.py", line 271, in get_cost
    config_id = self.config_ids[config]
KeyError: None
 ```

My current workaround that prevents this issue is 
```
accuracy = 2 #will only work when the train_size has 2 or less accuracy, it can be increased if you want to
n = 10**accuracy
tr_sz = float(str(train_size)[:2+accuracy])*n
test_size = (n-tr_sz)/n
```

Can this be fixed?

### PR DESCRIPTION
**Pull request by teresaconc**
Fix #538 Issue: Floating point issues when choosing different holdout set train sizes. 

Pass train_size and test_size as integers (number of samples) instead of passing floats (ratio). This way we prevent floating point errors in the multiplication (n_samples * train_size/test_size) that happened in some specific cases.

### CODE REVIEW

**Comment by teresaconc** [comment]
@mfeurer I commited some changes and added unit tests. 

I am uncertain in what's the supposed behaviour for multilabel classification tasks. 
In the original _train_evaluator.py_ for the non shuffle case we have ` tmp_train_size = int(np.floor(train_size * y.shape[0]))`, which means that if I have 10 samples with 2 labels, and since y is previously flatten through "ravel()" the tmp_train_size will be train_size*20 and not train_size*10. 

Should I keep this behaviour for both shuffle and not shuffle versions?
In other words, in this part of the code for MULTILABEL_CLASSIFICATION:
```
if shuffle:
      cv = ShuffleSplit(n_splits=1, train_size=train_size,
                                 test_size=test_size, random_state=1)
```
should I give the train_size has a ratio of the flatten or non flatten version?

**Comment by teresaconc** [comment]
> Looking at them they appear to work on rather simple data. Do they fail without your fix?

Yes, they fail without my fix. The issue was that for some train_sizes this line` test_size = 1- train_size ` was yielding extra decimal numbers, like test_size = 1 - 0.7 was resulting in 0.300000004, so the number of samples does not need to be particularly large for it to fail. I added two tests with sample size of 900, but let me know if you would like something more elaborated.

> Also, could you please remove the pytest_cache files from the pull request?

Yeah, sorry about that. Done

> y is only ravelled for non-multiclass problems.

`y = D.data['Y_train'].ravel()` is being used in get_splitter() for every type of problem. I moved it inside the 
```
if D.info['task'] in CLASSIFICATION_TASKS and \
                        D.info['task'] != MULTILABEL_CLASSIFICATION:
```

**Comment by teresaconc** [comment]
I changed the solution for a simpler version that it seems to be working ok too. Instead of providing the number of samples as the train and test size (integer) , I provide only the test_size (float). This way, sklearn won't do any problematic computation and train_size will always be the complement of test_size.

If you would rather go with the previous version I can revert the changes, but I think is more readable and simpler this way.

**Comment by mfeurer** [comment]
Thanks for your comments and code updates. The code looks much simpler now. Could you please fix the PEP8 issues, then I'd be happy to merge.

**Comment by teresaconc** [comment]
Done!

**Comment by mfeurer** [comment]
I just deleted the a c file which I believe sneaked in by accident.

Thanks a lot!
